scripts/compaction_bounds.zig

```diff
@@ -0,0 +1,128 @@
+
+    const transfers_per_second = 1_000_000;
+
+    const cases = .{
+        .ideal,
+        .actual,
+        .actual_with_small_level0,
+    };
+    inline for (cases) |case| {
+        std.debug.print("{}:\n", .{case});
+
+        const existing_transfer_counts = [_]usize{
+            0,
+            1_000_000,
+            10_000_000,
+            100_000_000,
+            1_000_000_000,
+        };
+        for (existing_transfer_counts) |existing_transfer_count| {
+            var bandwidth_bytes_per_second: usize = 0;
+
+            // Write bandwidth to WAL.
+            bandwidth_bytes_per_second += transfers_per_second * @sizeOf(Transfer);
+
+            const field_names = [_][]const u8{
+                "id",
+                "debit_account_id",
+                "credit_account_id",
+                "user_data",
+                "pending_id",
+                "timeout",
+                "ledger",
+                "code",
+                "amount",
+                "timestamp",
+                "balance",
+            };
+            inline for (field_names) |field_name| {
+                const Value = comptime if (std.meta.eql(field_name, "timestamp"))
+                    // Object tree.
+                    Transfer
+                else if (std.meta.eql(field_name, "balance"))
+                    // Proposed balance tree.
+                    struct {
+                        account_id: u128,
+                        transfer_id: u128,
+                        debits_pending: u64,
+                        debits_posted: u64,
+                        credits_pending: u64,
+                        credits_posted: u64,
+                    }
+                else
+                    // Secondary index.
+                    IndexCompositeKeyType(@TypeOf(@field(@as(Transfer, undefined), field_name)));
+
+                const values_per_table = @divTrunc(config.lsm_table_size_max, @sizeOf(Value));
+                const table_count = try std.math.divCeil(usize, existing_transfer_count, values_per_table);
+
+                var level_count: usize = 0;
+                {
+                    var table_count_remaining: usize = table_count;
+                    var tables_per_level: usize = config.lsm_growth_factor;
+                    while (table_count_remaining > 0) {
+                        level_count += 1;
+                        table_count_remaining = table_count_remaining -| tables_per_level;
+                        tables_per_level *= config.lsm_growth_factor;
+                    }
+                }
+
+                // Write bandwidth for flushing the immutable table:
+                bandwidth_bytes_per_second +=
+                    // Total size of all values
+                    transfers_per_second * @sizeOf(Value) *
+                    // All trees except the object tree might have to write tombstones
+                    @as(usize, if (std.meta.eql(field_name, "timestamp")) 1 else 2);
```

Comment:
Can you help me understand this better? Why would a tombstone cause two values to be written?

Reply:
Hmm, this is incorrect - we actually don't use any tombstones store balances in a separate tree. Will fix.